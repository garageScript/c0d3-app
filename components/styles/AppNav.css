.navbar__logo {
  font-family: 'PT Mono', sans-serif;
  font-weight: bold;
  font-size: 30px;
  margin-top: 0;
  margin-bottom: 0;
  color: rgb(84, 64, 216);
}

.navbar {
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  padding-top: 0.8rem;
}

.navbar__link {
  color: rgb(160,160,160);
  margin-left: 3em;
}

.navbar__link:hover {
  color: black;
}