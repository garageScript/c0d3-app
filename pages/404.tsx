import React from 'react'
import Error from '../components/Error'

export default function NotFound() {
  return <Error title="404 error" src="/404.png" message="Page not found" />
}
