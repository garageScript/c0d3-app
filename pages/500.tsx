import React from 'react'
import { WithLayout } from '../@types/page'
import Error, { StatusCode } from '../components/Error'
import { getLayout } from '../components/Layout'

const InternalServerError: React.FC & WithLayout = () => (
  <Error code={StatusCode.INTERNAL_SERVER_ERROR} />
)

InternalServerError.getLayout = getLayout

export default InternalServerError
