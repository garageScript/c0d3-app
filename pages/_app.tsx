import React, { useEffect } from 'react'
// TODO: type the posthog library
// @ts-ignore
import posthog from 'posthog-js'
import {
  ApolloProvider,
  ApolloClient,
  NormalizedCacheObject
} from '@apollo/client'
import { AppProps } from 'next/app'
import { MDXProvider } from '@mdx-js/react'
import Head from 'next/head'
import '../scss/index.scss'
import MDXcomponents from '../helpers/mdxComponents'
import { useApollo } from '../helpers/apolloClient'
import { ContextProvider } from '../helpers/globalContext'
import type { Page } from '../@types/page'
interface IProps extends AppProps {
  err: any
  apollo: ApolloClient<NormalizedCacheObject>
  Component: Page
}

function MyApp({ Component, pageProps, err }: IProps) {
  const getLayout = Component.getLayout || (page => page)
  const apolloClient = useApollo(pageProps)
  useEffect(() => {
    if (process.env.NODE_ENV === 'production' && process.env.POSTHOG_API_KEY) {
      posthog.init(process.env.POSTHOG_API_KEY, {
        api_host: 'https://app.posthog.com'
      })
    }
  }, [])
  return (
    <ApolloProvider client={apolloClient}>
      <MDXProvider components={MDXcomponents}>
        <ContextProvider>
          <Head>
            {/* <!-- Primary Meta Tags --> */}
            <meta name="title" content="C0D3" />
            <meta
              name="description"
              content="Learn the foundations to be a full stack software engineer. 100% free."
            />
            <link rel="shortcut icon" href="/favicon.ico" />

            {/* <!-- Open Graph / Facebook --> */}
            <meta property="og:type" content="website" />
            <meta property="og:url" content="https://www.c0d3.com/" />
            <meta
              property="og:title"
              content="C0D3 — Learn Javascript the old school way"
            />
            <meta
              property="og:description"
              content="Learn the foundations to be a full stack software engineer. 100% free."
            />
            <meta
              property="og:image"
              content="https://www.c0d3.com/assets/c0d3-meta.svg"
            />

            {/* <!-- Twitter --> */}
            <meta property="twitter:card" content="summary_large_image" />
            <meta property="twitter:url" content="https://www.c0d3.com/" />
            <meta
              property="twitter:title"
              content="C0D3 — Learn Javascript the old school way"
            />
            <meta
              property="twitter:description"
              content="Learn the foundations to be a full stack software engineer. 100% free."
            />
            <meta
              property="twitter:image"
              content="https://www.c0d3.com/assets/c0d3-meta.svg"
            />
          </Head>
          {getLayout(<Component {...pageProps} err={err} />, pageProps)}
        </ContextProvider>
      </MDXProvider>
    </ApolloProvider>
  )
}

export default MyApp
