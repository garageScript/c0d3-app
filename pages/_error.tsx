// From Next.JS example https://github.com/vercel/next.js/blob/canary/examples/with-sentry-simple/pages/_error.js
import React from 'react'
import * as Sentry from '@sentry/browser'
import { NextPageContext } from 'next'
import ErrorComponent, { StatusCode } from '../components/Error'

interface MyErrorContext extends NextPageContext {
  statusCode: number
  hasGetInitialPropsRun: boolean
}
const toStatusCode = (code: number): StatusCode => {
  if (code === 404) return StatusCode.NOT_FOUND
  return StatusCode.INTERNAL_SERVER_ERROR
}

const MyError = ({
  statusCode,
  hasGetInitialPropsRun,
  err
}: MyErrorContext) => {
  if (!hasGetInitialPropsRun && err) {
    // getInitialProps is not called in case of
    // https://github.com/vercel/next.js/issues/8592. As a workaround, we pass
    // err via _app.js so it can be captured
    Sentry.captureException(err)
  }

  return <ErrorComponent code={toStatusCode(statusCode)} />
}

MyError.getInitialProps = async ({ res, err, asPath }: MyErrorContext) => {
  const statusCode = res ? res.statusCode : err ? err.statusCode : 404
  const errorInitialProps = {
    statusCode,
    hasGetInitialPropsRun: false
  }

  // Workaround for https://github.com/vercel/next.js/issues/8592, mark when
  // getInitialProps has run
  errorInitialProps.hasGetInitialPropsRun = true

  // Running on the server, the response object (`res`) is available.
  //
  // Next.js will pass an err on the server if a page's data fetching methods
  // threw or returned a Promise that rejected
  //
  // Running on the client (browser), Next.js will provide an err if:
  //
  //  - a page's `getInitialProps` threw or returned a Promise that rejected
  //  - an exception was thrown somewhere in the React lifecycle (render,
  //    componentDidMount, etc) that was caught by Next.js's React Error
  //    Boundary. Read more about what types of exceptions are caught by Error
  //    Boundaries: https://reactjs.org/docs/error-boundaries.html
  if (err) {
    Sentry.captureException(err)
    return errorInitialProps
  }

  // If this point is reached, getInitialProps was called without any
  // information about what the error might be. This is unexpected and may
  // indicate a bug introduced in Next.js, so record it in Sentry
  Sentry.captureException(
    new Error(`_error.js getInitialProps missing data at path: ${asPath}`)
  )

  return errorInitialProps
}

export default MyError
