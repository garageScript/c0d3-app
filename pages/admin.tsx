import * as React from 'react'
import AppNav from '../components/AppNav'
import Footer from '../components/Footer'
import useSWR from 'swr'

import { useQuery } from '@apollo/react-hooks'
import { GET_LESSONS } from '../graphql/queries'

const SERVER_URL = process.env.SERVER_URL

export const fetcher = (url: string) =>
  fetch(url, { credentials: 'include' }).then(r => r.json())

const AdminPage: any = () => {
  const { data: userData, error: userError } = useSWR(
    `${SERVER_URL}/session`,
    fetcher
  )
  const { loading: lessonsLoading, data: lessonsData } = useQuery(GET_LESSONS)

  if ((!userData && !userError) || lessonsLoading) {
    return <h1>Loading...</h1>
  }

  console.log(userData, lessonsData)

  return (
    <>
      <AppNav
        firstName={userData.userInfo.name}
        username={userData.userInfo.username}
        loggedIn={true}
      />

      <div className="row mt-3">
        <h1 className="col-12">Lessons</h1>
        {lessonsData.lessons.map((lesson: any, i: number) => {
          let l: any = {}
          Object.assign(l, lesson)
          delete l.challenges
          delete l.__typename
          return (
            <div key={i} className="card shadow-sm col-3">
              <div className="card-body pt-5 pb-5">
                <h6 className="card-title font-weight-bold">
                  [ID:{l.id} | ORDER:{l.order}] - {l.title}
                </h6>
                <code>
                  <pre
                    style={{ width: '100%', whiteSpace: 'pre-line' }}
                    className="card-text"
                  >
                    {JSON.stringify(l)}
                  </pre>
                </code>
              </div>
            </div>
          )
        })}
      </div>

      {lessonsData.lessons.map((lesson: any, lidx: number) => {
        return (
          <div key={lidx} className="row mt-3">
            <h1 className="col-12">Lesson {lesson.order} Challenges</h1>
            {lesson.challenges.map((challenge: any, i: number) => {
              return (
                <div key={i} className="card shadow-sm col-3">
                  <div className="card-body pt-5 pb-5">
                    <h6 className="card-title font-weight-bold">
                      [ID:{challenge.id} | ORDER:{challenge.order}] -{' '}
                      {challenge.title}
                    </h6>
                    <code>
                      <pre
                        style={{ width: '100%', whiteSpace: 'pre-line' }}
                        className="card-text"
                      >
                        {JSON.stringify(challenge)}
                      </pre>
                    </code>
                  </div>
                </div>
              )
            })}
          </div>
        )
      })}

      <Footer footerType="py-5 bg-white text-muted" />
    </>
  )
}

export default AdminPage
