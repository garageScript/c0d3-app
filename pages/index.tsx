import React, { useEffect } from 'react'
import AppNav from '../components/AppNav'
import LandingPage from '../components/LandingPage'
import Footer from '../components/Footer'
import { useRouter } from 'next/router'
const IndexPage: React.FC<{}> = () => {
  const router = useRouter()
  useEffect(() => {
    localStorage.getItem('loggedIn') && router.push('/curriculum')
  }, [])
  return (
    <>
      <AppNav />
      <LandingPage />
      <Footer footerType="py-5 bg-white text-muted" />
    </>
  )
}

export default IndexPage
