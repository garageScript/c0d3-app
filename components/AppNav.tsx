import React, { useEffect, useState } from 'react'
import NavLink, { NavLinkProps } from './NavLink'
import { Button } from './theme/Button'
import { useRouter } from 'next/router'
import { DropdownMenu } from './DropdownMenu'
import { useLogoutMutation, useGetAppQuery, GetAppQuery } from '../graphql'
import _ from 'lodash'
import styles from '../scss/appNav.module.scss'

type AuthButtonProps = {
  initial: string
  username: string
}

type AuthLinkProps = {
  session: any
}

type NavItem = NavLinkProps & { name: string }

const dropdownMenuItems = [
  { title: 'Lessons', path: '/admin/lessons' },
  { title: 'Users', path: '/admin/users' },
  { title: 'Alerts', path: '/admin/alerts' }
]
const navItems: NavItem[] = [
  { path: '/curriculum', name: 'Curriculum' },
  {
    path: 'https://github.com/garageScript/c0d3-app',
    name: 'Repo',
    external: true
  },
  {
    path: '/docs/learn',
    name: 'Learn'
  },
  {
    path: 'https://discord.gg/c0d3',
    name: 'Community',
    external: true
  }
]

const NavBar: React.FC<AuthLinkProps> = ({ session }) => {
  const router = useRouter()
  const isAdmin = _.get(session, 'user.isAdmin', false) as boolean
  const location = '/' + router.asPath.split('/')[1]

  return (
    <div className="navbar-nav collapse navbar-collapse">
      {navItems.map(button => (
        <NavLink
          {...button}
          className={`${styles['nav-item']} nav-link`}
          key={button.name}
          activePath={location === button.path}
        >
          {button.name}
        </NavLink>
      ))}
      {isAdmin && <DropdownMenu title="Admin" items={dropdownMenuItems} />}
    </div>
  )
}

const AuthButton: React.FC<AuthButtonProps> = ({ initial, username }) => {
  const router = useRouter()
  const [logoutUser] = useLogoutMutation({
    update(cache) {
      cache.modify({
        fields: {
          session() {
            return { lessonStatus: [] }
          }
        },
        broadcast: false
      })
    },
    onCompleted: () => {
      window.localStorage.removeItem('loggedIn')
      router.push('/')
    }
  })
  return (
    <div className="d-flex">
      <NavLink
        path="/profile/[username]"
        as={`/profile/${username}`}
        className="btn btn-secondary border overflow-hidden p-2 text-truncate"
      >
        {`${initial} ${username}`}
      </NavLink>

      <Button border ml="2" onClick={logoutUser}>
        Logout
      </Button>
    </div>
  )
}

const UnAuthButton = () => (
  <div>
    <NavLink path="/login" className="btn btn-secondary border mr-3">
      Login
    </NavLink>
    <NavLink path="/signup" className="btn btn-secondary border mr-3">
      Signup
    </NavLink>
  </div>
)

const AppNav: React.FC<{}> = () => {
  const [session, setSession] = useState<GetAppQuery['session']>({
    lessonStatus: []
  })
  const { data } = useGetAppQuery()
  useEffect(() => {
    if (data && data.session) {
      setSession(data.session)
    }
  }, [data])
  const renderButtons = () => {
    if (!session || _.get(session, 'user.username', null) === null) {
      return <UnAuthButton />
    }

    const initial = ''
    // TODO: replace with typing
    const username = _.get(session, 'user.username', '')

    return <AuthButton username={username} initial={initial} />
  }

  return (
    <nav className="navbar navbar-expand-lg navbar-light justify-content-between bg-white">
      <div className="container">
        <NavLink
          path="/"
          className={`${styles['navbar-brand']} text-primary font-weight-bold`}
        >
          C0D3
        </NavLink>
        <div id="navbarNav">
          <div className="navbar-nav collapse navbar-collapse">
            <NavBar session={session} />
          </div>
        </div>
        {renderButtons()}
      </div>
    </nav>
  )
}

export default AppNav
