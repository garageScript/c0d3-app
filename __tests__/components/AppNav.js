import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import { AuthUserContext } from '../../pages/_app'
import AppNav from '../../components/AppNav'

describe('AppNav component', () => {
  test('Should render appropriately when user is logged in', () => {
    const user = { user: { username: 'test' } }

    const tree = (
      <AuthUserContext.Provider value={user}>
        <AppNav />
      </AuthUserContext.Provider>
    )

    const { container } = render(tree)
    expect(container).toMatchSnapshot()
  })

  test('Should render appropriately when user is not logged in', () => {
    const tree = (
      <AuthUserContext.Provider value={false}>
        <AppNav />
      </AuthUserContext.Provider>
    )

    const { container } = render(tree)
    expect(container).toMatchSnapshot()
  })
})
