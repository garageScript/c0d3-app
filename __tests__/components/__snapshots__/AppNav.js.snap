// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppNav component Should render appropriately when user is logged in 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/home"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum 
              <span
                class="sr-only"
              >
                (current)
              </span>
            </a>
            <a
              class="nav-item nav-link"
              href="#"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="#"
            >
              Journey
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
          </div>
        </div>
      </div>
      <div>
        <button
          class="btn border btn-secondary overflow-hidden p-2 text-truncate"
        />
        <button
          class="btn border btn-secondary ml-2"
        >
          Logout
        </button>
      </div>
    </div>
  </nav>
</div>
`;

exports[`AppNav component Should render appropriately when user is not logged in 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/home"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/"
            >
              Home 
              <span
                class="sr-only"
              >
                (current)
              </span>
            </a>
            <a
              class="nav-item nav-link"
              href="/#learning"
            >
              Learning Process
            </a>
            <a
              class="nav-item nav-link"
              href="https://c0d3.com/book"
              rel="noopener noreferrer"
              target="_blank"
            >
              Resources
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
</div>
`;
